{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/scratch/l_singh/hdb/project_dir/histonedb/CURATED_SET\n"
     ]
    }
   ],
   "source": [
    "%cd '..'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import inspect\n",
    "import os\n",
    "import re\n",
    "import sys\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import Entrez, SeqIO\n",
    "from curated_set_services import CuratedSet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((561, 16),\n",
       " Index(['accession', 'type', 'variant_group', 'variant', 'doublet', 'gi',\n",
       "        'ncbi_gene_id', 'hgnc_gene_name', 'taxonomy_id', 'organism', 'phylum',\n",
       "        'class', 'taxonomy_group', 'info', 'references', 'sequence'],\n",
       "       dtype='object'))"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load data from histones.csv\n",
    "curated_set = CuratedSet()\n",
    "cs = curated_set\n",
    "\n",
    "cs.data.shape, cs.data.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "my_columns = cs.data.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cs.data = cs.data[cs.data[\"variant_group\"] != \"gH2B\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(cs.has_duplicates())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CUT18450.1</th>\n",
       "      <td>CUT18450.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.S</td>\n",
       "      <td>H2B.S</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1473204</td>\n",
       "      <td>Lilium davidii var. unicolor</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>32716939</td>\n",
       "      <td>MAPKKKPSKLVGTVTKTRKVTETQTLKVSLTKGLKPEDQQTTTNKF...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             accession type variant_group variant doublet gi ncbi_gene_id  \\\n",
       "CUT18450.1  CUT18450.1  H2B         H2B.S   H2B.S                           \n",
       "\n",
       "           hgnc_gene_name taxonomy_id                      organism  \\\n",
       "CUT18450.1                    1473204  Lilium davidii var. unicolor   \n",
       "\n",
       "                  phylum          class taxonomy_group info references  \\\n",
       "CUT18450.1  Streptophyta  Magnoliopsida                       32716939   \n",
       "\n",
       "                                                     sequence  \n",
       "CUT18450.1  MAPKKKPSKLVGTVTKTRKVTETQTLKVSLTKGLKPEDQQTTTNKF...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data[cs.data[\"accession\"] == \"CUT18450.1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NP_505463.1</th>\n",
       "      <td>NP_505463.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>cH2A_(Animals)</td>\n",
       "      <td></td>\n",
       "      <td>17562014</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>6239</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>Nematoda</td>\n",
       "      <td>Chromadorea</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26989147 22650316(?)</td>\n",
       "      <td>MSGRGKGGKAKTGGKAKSRSSRAGLQFPVGRLHRILRKGNYAQRVG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EEC09557.1</th>\n",
       "      <td>EEC09557.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>cH2A_(Animals)</td>\n",
       "      <td></td>\n",
       "      <td>215500063</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>6945</td>\n",
       "      <td>Ixodes scapularis</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Arachnida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26989147 22650316(?)</td>\n",
       "      <td>MSGRGKGGKVKGKSKTRSSRAGLQFPVGRIHRLLRKGNYAERVGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_724343.1</th>\n",
       "      <td>NP_724343.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>cH2A_(Animals)</td>\n",
       "      <td></td>\n",
       "      <td>24585673</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26989147 22650316(?)</td>\n",
       "      <td>MSGRGKGGKVKGKAKSRSNRAGLQFPVGRIHRLLRKGNYAERVGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XP_001119899.1</th>\n",
       "      <td>XP_001119899.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>cH2A_(Animals)</td>\n",
       "      <td></td>\n",
       "      <td>110764935</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7460</td>\n",
       "      <td>Apis mellifera</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26989147 22650316(?)</td>\n",
       "      <td>MSGRGKGGKAKAKAKSRSNRAGLQFPVGRIHRLLRKGNYAERVGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EDO48405.1</th>\n",
       "      <td>EDO48405.1</td>\n",
       "      <td>H2A</td>\n",
       "      <td>cH2A</td>\n",
       "      <td>cH2A_(Animals)</td>\n",
       "      <td></td>\n",
       "      <td>156227602</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>45351</td>\n",
       "      <td>Nematostella vectensis</td>\n",
       "      <td>Cnidaria</td>\n",
       "      <td>Anthozoa</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26989147 22650316(?)</td>\n",
       "      <td>MSGRGKGKAKGTKSKTRSSRAGLQFPVGRIHRHLRKGNYAERVGAG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     accession type variant_group         variant doublet  \\\n",
       "NP_505463.1        NP_505463.1  H2A          cH2A  cH2A_(Animals)           \n",
       "EEC09557.1          EEC09557.1  H2A          cH2A  cH2A_(Animals)           \n",
       "NP_724343.1        NP_724343.1  H2A          cH2A  cH2A_(Animals)           \n",
       "XP_001119899.1  XP_001119899.1  H2A          cH2A  cH2A_(Animals)           \n",
       "EDO48405.1          EDO48405.1  H2A          cH2A  cH2A_(Animals)           \n",
       "\n",
       "                       gi ncbi_gene_id hgnc_gene_name taxonomy_id  \\\n",
       "NP_505463.1      17562014                                    6239   \n",
       "EEC09557.1      215500063                                    6945   \n",
       "NP_724343.1      24585673                                    7227   \n",
       "XP_001119899.1  110764935                                    7460   \n",
       "EDO48405.1      156227602                                   45351   \n",
       "\n",
       "                               organism      phylum        class  \\\n",
       "NP_505463.1      Caenorhabditis elegans    Nematoda  Chromadorea   \n",
       "EEC09557.1            Ixodes scapularis  Arthropoda    Arachnida   \n",
       "NP_724343.1     Drosophila melanogaster  Arthropoda      Insecta   \n",
       "XP_001119899.1           Apis mellifera  Arthropoda      Insecta   \n",
       "EDO48405.1       Nematostella vectensis    Cnidaria     Anthozoa   \n",
       "\n",
       "               taxonomy_group info            references  \\\n",
       "NP_505463.1                         26989147 22650316(?)   \n",
       "EEC09557.1                          26989147 22650316(?)   \n",
       "NP_724343.1                         26989147 22650316(?)   \n",
       "XP_001119899.1                      26989147 22650316(?)   \n",
       "EDO48405.1                          26989147 22650316(?)   \n",
       "\n",
       "                                                         sequence  \n",
       "NP_505463.1     MSGRGKGGKAKTGGKAKSRSSRAGLQFPVGRLHRILRKGNYAQRVG...  \n",
       "EEC09557.1      MSGRGKGGKVKGKSKTRSSRAGLQFPVGRIHRLLRKGNYAERVGAG...  \n",
       "NP_724343.1     MSGRGKGGKVKGKAKSRSNRAGLQFPVGRIHRLLRKGNYAERVGAG...  \n",
       "XP_001119899.1  MSGRGKGGKAKAKAKSRSNRAGLQFPVGRIHRLLRKGNYAERVGAG...  \n",
       "EDO48405.1      MSGRGKGKAKGTKSKTRSSRAGLQFPVGRIHRHLRKGNYAERVGAG...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(548, 16)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CUT18450.1</th>\n",
       "      <td>CUT18450.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>H2B.S</td>\n",
       "      <td>H2B.S</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1473204</td>\n",
       "      <td>Lilium davidii var. unicolor</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>32716939</td>\n",
       "      <td>MAPKKKPSKLVGTVTKTRKVTETQTLKVSLTKGLKPEDQQTTTNKF...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             accession type variant_group variant doublet gi ncbi_gene_id  \\\n",
       "CUT18450.1  CUT18450.1  H2B         H2B.S   H2B.S                           \n",
       "\n",
       "           hgnc_gene_name taxonomy_id                      organism  \\\n",
       "CUT18450.1                    1473204  Lilium davidii var. unicolor   \n",
       "\n",
       "                  phylum          class taxonomy_group info references  \\\n",
       "CUT18450.1  Streptophyta  Magnoliopsida                       32716939   \n",
       "\n",
       "                                                     sequence  \n",
       "CUT18450.1  MAPKKKPSKLVGTVTKTRKVTETQTLKVSLTKGLKPEDQQTTTNKF...  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g_h2b_accessions = [\n",
    "    \"CUT18449.1\",\n",
    "    \"CUT18450.1\",\n",
    "    \"CUT18445.1\",\n",
    "    \"CUT18446.1\",\n",
    "    \"CUT18447.1\",\n",
    "    \"CUT18448.1\",\n",
    "    \"CUT18451.1\",\n",
    "    \"CUT18452.1\",\n",
    "]\n",
    "cs.data.loc[cs.data.index.intersection(g_h2b_accessions), :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((8, 5),\n",
       " Index(['accession', 'type', 'variant_group', 'variant', 'references'], dtype='object'))"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(\n",
    "    {\n",
    "        \"accession\": g_h2b_accessions,\n",
    "        \"type\": [\"H2B\"] * len(g_h2b_accessions),\n",
    "        \"variant_group\": [\"gH2B\"] * len(g_h2b_accessions),\n",
    "        \"variant\": [\"gH2B\"] * len(g_h2b_accessions),\n",
    "        \"references\": [\"26846354\"] * len(g_h2b_accessions),\n",
    "    }\n",
    ")\n",
    "df.index = df.accession\n",
    "df.at[cs.data.index.intersection(g_h2b_accessions), \"references\"] = (\n",
    "    \"26846354 \"\n",
    "    + cs.data.loc[cs.data.index.intersection(g_h2b_accessions), \"references\"]\n",
    ")\n",
    "df.shape, df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>references</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CUT18449.1</th>\n",
       "      <td>CUT18449.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>26846354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18450.1</th>\n",
       "      <td>CUT18450.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>26846354 32716939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18445.1</th>\n",
       "      <td>CUT18445.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>26846354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18446.1</th>\n",
       "      <td>CUT18446.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>26846354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18447.1</th>\n",
       "      <td>CUT18447.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>26846354</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             accession type variant_group variant         references\n",
       "accession                                                           \n",
       "CUT18449.1  CUT18449.1  H2B          gH2B    gH2B           26846354\n",
       "CUT18450.1  CUT18450.1  H2B          gH2B    gH2B  26846354 32716939\n",
       "CUT18445.1  CUT18445.1  H2B          gH2B    gH2B           26846354\n",
       "CUT18446.1  CUT18446.1  H2B          gH2B    gH2B           26846354\n",
       "CUT18447.1  CUT18447.1  H2B          gH2B    gH2B           26846354"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(547, 16)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data = cs.data[cs.data['accession'] != cs.data.index.intersection(g_h2b_accessions).values[0]]\n",
    "cs.data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, type, variant_group, variant, doublet, gi, ncbi_gene_id, hgnc_gene_name, taxonomy_id, organism, phylum, class, taxonomy_group, info, references, sequence]\n",
       "Index: []"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data[cs.data[\"accession\"] == \"CUT18450.1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cp histones.csv backups/histones.csv-Feb0624193432\n",
      "Previous data backuped to backups/histones.csv-Feb0624193432\n",
      "Results saved to histones.csv\n"
     ]
    }
   ],
   "source": [
    "cs.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((555, 16), [])"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data = pd.concat([cs.data, df]).fillna(\"\")\n",
    "cs.data.shape, list(cs.has_duplicates())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CUT18446.1</th>\n",
       "      <td>CUT18446.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18447.1</th>\n",
       "      <td>CUT18447.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18448.1</th>\n",
       "      <td>CUT18448.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18451.1</th>\n",
       "      <td>CUT18451.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18452.1</th>\n",
       "      <td>CUT18452.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             accession type variant_group variant doublet gi ncbi_gene_id  \\\n",
       "CUT18446.1  CUT18446.1  H2B          gH2B    gH2B                           \n",
       "CUT18447.1  CUT18447.1  H2B          gH2B    gH2B                           \n",
       "CUT18448.1  CUT18448.1  H2B          gH2B    gH2B                           \n",
       "CUT18451.1  CUT18451.1  H2B          gH2B    gH2B                           \n",
       "CUT18452.1  CUT18452.1  H2B          gH2B    gH2B                           \n",
       "\n",
       "           hgnc_gene_name taxonomy_id organism phylum class taxonomy_group  \\\n",
       "CUT18446.1                                                                   \n",
       "CUT18447.1                                                                   \n",
       "CUT18448.1                                                                   \n",
       "CUT18451.1                                                                   \n",
       "CUT18452.1                                                                   \n",
       "\n",
       "           info references sequence  \n",
       "CUT18446.1        26846354           \n",
       "CUT18447.1        26846354           \n",
       "CUT18448.1        26846354           \n",
       "CUT18451.1        26846354           \n",
       "CUT18452.1        26846354           "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8, 16)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data[cs.data[\"variant\"] == \"gH2B\"].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/l_singh/.conda/envs/histdb_env/lib/python3.8/site-packages/Bio/Entrez/__init__.py:658: UserWarning: \n",
      "Email address is not specified.\n",
      "\n",
      "To make use of NCBI's E-utilities, NCBI requires you to specify your\n",
      "email address with each request.  As an example, if your email address\n",
      "is A.N.Other@example.com, you can specify it as follows:\n",
      "   from Bio import Entrez\n",
      "   Entrez.email = 'A.N.Other@example.com'\n",
      "In case of excessive usage of the E-utilities, NCBI will attempt to contact\n",
      "a user at the email address provided before blocking access to the\n",
      "E-utilities.\n",
      "  warnings.warn(\n",
      "/home/l_singh/.conda/envs/histdb_env/lib/python3.8/site-packages/Bio/GenBank/__init__.py:1143: BiopythonParserWarning: Dropping bond qualifier in feature location\n",
      "  warnings.warn(\n",
      "/mnt/scratch/l_singh/hdb/project_dir/histonedb/CURATED_SET/curated_set_services.py:267: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  updating_data['accession'] = new_accessions\n"
     ]
    }
   ],
   "source": [
    "cs.update_accession_version()\n",
    "cs.data = cs.data.set_index(cs.data.accession.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8, 16)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data[cs.data[\"variant\"] == \"gH2B\"].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched taxid from NCBI 1473204\n",
      "Fetched taxid from NCBI 1473204\n",
      "Fetched taxid from NCBI 1473204\n",
      "Unexpected error: <class 'urllib.error.HTTPError'>, Retrying, attempt 0\n",
      "Fetched taxid from NCBI 1473204\n",
      "Fetched taxid from NCBI 1473204\n",
      "Fetched taxid from NCBI 1473204\n",
      "Fetched taxid from NCBI 1473204\n",
      "Fetched taxid from NCBI 1473204\n",
      " changes to 1473204\n",
      " changes to Lilium davidii var. unicolor\n",
      " changes to Streptophyta\n",
      " changes to Magnoliopsida\n",
      " changes to 1473204\n",
      " changes to Lilium davidii var. unicolor\n",
      " changes to Streptophyta\n",
      " changes to Magnoliopsida\n",
      " changes to 1473204\n",
      " changes to Lilium davidii var. unicolor\n",
      " changes to Streptophyta\n",
      " changes to Magnoliopsida\n",
      " changes to 1473204\n",
      " changes to Lilium davidii var. unicolor\n",
      " changes to Streptophyta\n",
      " changes to Magnoliopsida\n",
      " changes to 1473204\n",
      " changes to Lilium davidii var. unicolor\n",
      " changes to Streptophyta\n",
      " changes to Magnoliopsida\n",
      " changes to 1473204\n",
      " changes to Lilium davidii var. unicolor\n",
      " changes to Streptophyta\n",
      " changes to Magnoliopsida\n",
      " changes to 1473204\n",
      " changes to Lilium davidii var. unicolor\n",
      " changes to Streptophyta\n",
      " changes to Magnoliopsida\n",
      " changes to 1473204\n",
      " changes to Lilium davidii var. unicolor\n",
      " changes to Streptophyta\n",
      " changes to Magnoliopsida\n"
     ]
    }
   ],
   "source": [
    "curated_set.update_taxids(blank_data=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8, 16)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data[cs.data[\"variant\"] == \"gH2B\"].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CUT18449.1</th>\n",
       "      <td>CUT18449.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1473204</td>\n",
       "      <td>Lilium davidii var. unicolor</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18450.1</th>\n",
       "      <td>CUT18450.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1473204</td>\n",
       "      <td>Lilium davidii var. unicolor</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354 32716939</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18445.1</th>\n",
       "      <td>CUT18445.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1473204</td>\n",
       "      <td>Lilium davidii var. unicolor</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18446.1</th>\n",
       "      <td>CUT18446.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1473204</td>\n",
       "      <td>Lilium davidii var. unicolor</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18447.1</th>\n",
       "      <td>CUT18447.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1473204</td>\n",
       "      <td>Lilium davidii var. unicolor</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18448.1</th>\n",
       "      <td>CUT18448.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1473204</td>\n",
       "      <td>Lilium davidii var. unicolor</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18451.1</th>\n",
       "      <td>CUT18451.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1473204</td>\n",
       "      <td>Lilium davidii var. unicolor</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18452.1</th>\n",
       "      <td>CUT18452.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1473204</td>\n",
       "      <td>Lilium davidii var. unicolor</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             accession type variant_group variant doublet gi ncbi_gene_id  \\\n",
       "CUT18449.1  CUT18449.1  H2B          gH2B    gH2B                           \n",
       "CUT18450.1  CUT18450.1  H2B          gH2B    gH2B                           \n",
       "CUT18445.1  CUT18445.1  H2B          gH2B    gH2B                           \n",
       "CUT18446.1  CUT18446.1  H2B          gH2B    gH2B                           \n",
       "CUT18447.1  CUT18447.1  H2B          gH2B    gH2B                           \n",
       "CUT18448.1  CUT18448.1  H2B          gH2B    gH2B                           \n",
       "CUT18451.1  CUT18451.1  H2B          gH2B    gH2B                           \n",
       "CUT18452.1  CUT18452.1  H2B          gH2B    gH2B                           \n",
       "\n",
       "           hgnc_gene_name taxonomy_id                      organism  \\\n",
       "CUT18449.1                    1473204  Lilium davidii var. unicolor   \n",
       "CUT18450.1                    1473204  Lilium davidii var. unicolor   \n",
       "CUT18445.1                    1473204  Lilium davidii var. unicolor   \n",
       "CUT18446.1                    1473204  Lilium davidii var. unicolor   \n",
       "CUT18447.1                    1473204  Lilium davidii var. unicolor   \n",
       "CUT18448.1                    1473204  Lilium davidii var. unicolor   \n",
       "CUT18451.1                    1473204  Lilium davidii var. unicolor   \n",
       "CUT18452.1                    1473204  Lilium davidii var. unicolor   \n",
       "\n",
       "                  phylum          class taxonomy_group info  \\\n",
       "CUT18449.1  Streptophyta  Magnoliopsida                       \n",
       "CUT18450.1  Streptophyta  Magnoliopsida                       \n",
       "CUT18445.1  Streptophyta  Magnoliopsida                       \n",
       "CUT18446.1  Streptophyta  Magnoliopsida                       \n",
       "CUT18447.1  Streptophyta  Magnoliopsida                       \n",
       "CUT18448.1  Streptophyta  Magnoliopsida                       \n",
       "CUT18451.1  Streptophyta  Magnoliopsida                       \n",
       "CUT18452.1  Streptophyta  Magnoliopsida                       \n",
       "\n",
       "                   references sequence  \n",
       "CUT18449.1           26846354           \n",
       "CUT18450.1  26846354 32716939           \n",
       "CUT18445.1           26846354           \n",
       "CUT18446.1           26846354           \n",
       "CUT18447.1           26846354           \n",
       "CUT18448.1           26846354           \n",
       "CUT18451.1           26846354           \n",
       "CUT18452.1           26846354           "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data[cs.data[\"variant\"] == \"gH2B\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True])"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_columns == cs.data.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Added sequence with accession CUT18452.1\n",
      "Added sequence with accession CUT18445.1\n",
      "Added sequence with accession CUT18448.1\n",
      "Added sequence with accession CUT18446.1\n",
      "Added sequence with accession CUT18447.1\n",
      "Added sequence with accession CUT18451.1\n",
      "Added sequence with accession CUT18450.1\n",
      "Added sequence with accession CUT18449.1\n",
      "cp histones.csv backups/histones.csv-Feb0624193542\n",
      "Previous data backuped to backups/histones.csv-Feb0624193542\n",
      "Results saved to histones.csv\n"
     ]
    }
   ],
   "source": [
    "cs.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python 3.8.1\n"
     ]
    }
   ],
   "source": [
    "!python --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Updating sequences for gH2B (added after adding H2B.Z 6-Mar-2024)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CUT18449.1</th>\n",
       "      <td>CUT18449.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1473204</td>\n",
       "      <td>Lilium davidii var. unicolor</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18450.1</th>\n",
       "      <td>CUT18450.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1473204</td>\n",
       "      <td>Lilium davidii var. unicolor</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354 32716939</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18445.1</th>\n",
       "      <td>CUT18445.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1473204</td>\n",
       "      <td>Lilium davidii var. unicolor</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18446.1</th>\n",
       "      <td>CUT18446.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1473204</td>\n",
       "      <td>Lilium davidii var. unicolor</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18447.1</th>\n",
       "      <td>CUT18447.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1473204</td>\n",
       "      <td>Lilium davidii var. unicolor</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18448.1</th>\n",
       "      <td>CUT18448.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1473204</td>\n",
       "      <td>Lilium davidii var. unicolor</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18451.1</th>\n",
       "      <td>CUT18451.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1473204</td>\n",
       "      <td>Lilium davidii var. unicolor</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18452.1</th>\n",
       "      <td>CUT18452.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1473204</td>\n",
       "      <td>Lilium davidii var. unicolor</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             accession type variant_group variant doublet gi ncbi_gene_id  \\\n",
       "CUT18449.1  CUT18449.1  H2B          gH2B    gH2B                           \n",
       "CUT18450.1  CUT18450.1  H2B          gH2B    gH2B                           \n",
       "CUT18445.1  CUT18445.1  H2B          gH2B    gH2B                           \n",
       "CUT18446.1  CUT18446.1  H2B          gH2B    gH2B                           \n",
       "CUT18447.1  CUT18447.1  H2B          gH2B    gH2B                           \n",
       "CUT18448.1  CUT18448.1  H2B          gH2B    gH2B                           \n",
       "CUT18451.1  CUT18451.1  H2B          gH2B    gH2B                           \n",
       "CUT18452.1  CUT18452.1  H2B          gH2B    gH2B                           \n",
       "\n",
       "           hgnc_gene_name taxonomy_id                      organism  \\\n",
       "CUT18449.1                    1473204  Lilium davidii var. unicolor   \n",
       "CUT18450.1                    1473204  Lilium davidii var. unicolor   \n",
       "CUT18445.1                    1473204  Lilium davidii var. unicolor   \n",
       "CUT18446.1                    1473204  Lilium davidii var. unicolor   \n",
       "CUT18447.1                    1473204  Lilium davidii var. unicolor   \n",
       "CUT18448.1                    1473204  Lilium davidii var. unicolor   \n",
       "CUT18451.1                    1473204  Lilium davidii var. unicolor   \n",
       "CUT18452.1                    1473204  Lilium davidii var. unicolor   \n",
       "\n",
       "                  phylum          class taxonomy_group info  \\\n",
       "CUT18449.1  Streptophyta  Magnoliopsida                       \n",
       "CUT18450.1  Streptophyta  Magnoliopsida                       \n",
       "CUT18445.1  Streptophyta  Magnoliopsida                       \n",
       "CUT18446.1  Streptophyta  Magnoliopsida                       \n",
       "CUT18447.1  Streptophyta  Magnoliopsida                       \n",
       "CUT18448.1  Streptophyta  Magnoliopsida                       \n",
       "CUT18451.1  Streptophyta  Magnoliopsida                       \n",
       "CUT18452.1  Streptophyta  Magnoliopsida                       \n",
       "\n",
       "                   references sequence  \n",
       "CUT18449.1           26846354           \n",
       "CUT18450.1  26846354 32716939           \n",
       "CUT18445.1           26846354           \n",
       "CUT18446.1           26846354           \n",
       "CUT18447.1           26846354           \n",
       "CUT18448.1           26846354           \n",
       "CUT18451.1           26846354           \n",
       "CUT18452.1           26846354           "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data[cs.data['sequence'] == '']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading FASTA SeqRecords by ACCESSIONs from NCBI\n",
      "Fetching 8 seqs\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/l_singh/.conda/envs/histdb_env/lib/python3.8/site-packages/Bio/Entrez/__init__.py:658: UserWarning: \n",
      "Email address is not specified.\n",
      "\n",
      "To make use of NCBI's E-utilities, NCBI requires you to specify your\n",
      "email address with each request.  As an example, if your email address\n",
      "is A.N.Other@example.com, you can specify it as follows:\n",
      "   from Bio import Entrez\n",
      "   Entrez.email = 'A.N.Other@example.com'\n",
      "In case of excessive usage of the E-utilities, NCBI will attempt to contact\n",
      "a user at the email address provided before blocking access to the\n",
      "E-utilities.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence for CUT18449.1 changes from  to MPPRRKKTAAGAAAGGKAAAAAVGKAGFMPPKKPKKGKKKTPIMRYKRYIYKVLKQVRPELGISEKSTMIMNNFVAHNFQNIAKEASILAYYSKKRTITVDELKAAVAMVLPNLLADYANRDGEKAVSNFEGEASAKKSQGRKRGRGQQA\n",
      "Sequence for CUT18450.1 changes from  to MAPKKKPSKLVGTVTKTRKVTETQTLKVSLTKGLKPEDQQTTTNKFEVSVTGKQSKTQPLIVSTNTNLVPKKEKEESPTTTLMVKKKRKNRKAGGEYKRYVYMVLKTVHPDMTVSSKAMMVMEGMMQDMFERLVTEAVRLVQYMKKATLTCREIQTAVMLVLPGELGKHAVSEGAKAITNYMAAVGSGNGGAA\n",
      "Sequence for CUT18445.1 changes from  to MAPKSEKKPAEKKPVAEKPAAEEEKKSAPAPAAAEKKPAEKKPKAGKKLPASKGKEGEKKK\n",
      "Sequence for CUT18446.1 changes from  to AEKKPKAGKKVPASKEGEKKKKRSKKSVETYKIYIFKVLKQVHPDIGISSKAMGIMNSFINDLFEKMAQESARLARYNKKNTITSREIQTSVRLVLPGELAKHAVSEGTKAVTKFTSA\n",
      "Sequence for CUT18447.1 changes from  to MAPKAEKKPAAKKPAATPPPEEEKEVVPPPPAEKKPKAGKKLPAAKEGDAKKKKKAKKSIETYKIYIFKVLKQVHPDIGISSKAMGIMNSFINDIFEKLAQESARLARYNKKPTITSREIQTSVRLVLPGELAKHAVSEGTKAVTKFTSA\n",
      "Sequence for CUT18448.1 changes from  to AATPPPEEEKEVVPPPAEKKPAEKKPKAGKKLPASKEGDAKKKKKSKKSIETYKIYIFKVLKQVHPDIGISSKAMGIMNSFINDIFEKLAQESARLARYNKKPTITSREIQTSVRLVLPGELAKHAVSEGTKAVTKFTSA\n",
      "Sequence for CUT18451.1 changes from  to MAPKAEKKPAAKKPAATPPPEEEKEVVPPPPAEKKPKAGKKLPAAKEGDAKKKKKAKKSIETYKIYIFKVLKQVHPDIGISSKAMGIMNSFINDIFEKLAQESARLARYNKKPTITSREIQTSVRLVLPGELAKHAVSEGTKAVTKFTSG\n",
      "Sequence for CUT18452.1 changes from  to MAPKSEKKPAEKKPVAEKPAAEEEKKAAPAAAPAEKKAAEKKPKA\n",
      "Sequences updated: 8\n"
     ]
    }
   ],
   "source": [
    "cs.update_sequence(blank_data=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>type</th>\n",
       "      <th>variant_group</th>\n",
       "      <th>variant</th>\n",
       "      <th>doublet</th>\n",
       "      <th>gi</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>hgnc_gene_name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>taxonomy_group</th>\n",
       "      <th>info</th>\n",
       "      <th>references</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CUT18449.1</th>\n",
       "      <td>CUT18449.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1473204</td>\n",
       "      <td>Lilium davidii var. unicolor</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354</td>\n",
       "      <td>MPPRRKKTAAGAAAGGKAAAAAVGKAGFMPPKKPKKGKKKTPIMRY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18450.1</th>\n",
       "      <td>CUT18450.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1473204</td>\n",
       "      <td>Lilium davidii var. unicolor</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354 32716939</td>\n",
       "      <td>MAPKKKPSKLVGTVTKTRKVTETQTLKVSLTKGLKPEDQQTTTNKF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18445.1</th>\n",
       "      <td>CUT18445.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1473204</td>\n",
       "      <td>Lilium davidii var. unicolor</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354</td>\n",
       "      <td>MAPKSEKKPAEKKPVAEKPAAEEEKKSAPAPAAAEKKPAEKKPKAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18446.1</th>\n",
       "      <td>CUT18446.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1473204</td>\n",
       "      <td>Lilium davidii var. unicolor</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354</td>\n",
       "      <td>AEKKPKAGKKVPASKEGEKKKKRSKKSVETYKIYIFKVLKQVHPDI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18447.1</th>\n",
       "      <td>CUT18447.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1473204</td>\n",
       "      <td>Lilium davidii var. unicolor</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354</td>\n",
       "      <td>MAPKAEKKPAAKKPAATPPPEEEKEVVPPPPAEKKPKAGKKLPAAK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18448.1</th>\n",
       "      <td>CUT18448.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1473204</td>\n",
       "      <td>Lilium davidii var. unicolor</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354</td>\n",
       "      <td>AATPPPEEEKEVVPPPAEKKPAEKKPKAGKKLPASKEGDAKKKKKS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18451.1</th>\n",
       "      <td>CUT18451.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1473204</td>\n",
       "      <td>Lilium davidii var. unicolor</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354</td>\n",
       "      <td>MAPKAEKKPAAKKPAATPPPEEEKEVVPPPPAEKKPKAGKKLPAAK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CUT18452.1</th>\n",
       "      <td>CUT18452.1</td>\n",
       "      <td>H2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td>gH2B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1473204</td>\n",
       "      <td>Lilium davidii var. unicolor</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26846354</td>\n",
       "      <td>MAPKSEKKPAEKKPVAEKPAAEEEKKAAPAAAPAEKKAAEKKPKA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             accession type variant_group variant doublet gi ncbi_gene_id  \\\n",
       "CUT18449.1  CUT18449.1  H2B          gH2B    gH2B                           \n",
       "CUT18450.1  CUT18450.1  H2B          gH2B    gH2B                           \n",
       "CUT18445.1  CUT18445.1  H2B          gH2B    gH2B                           \n",
       "CUT18446.1  CUT18446.1  H2B          gH2B    gH2B                           \n",
       "CUT18447.1  CUT18447.1  H2B          gH2B    gH2B                           \n",
       "CUT18448.1  CUT18448.1  H2B          gH2B    gH2B                           \n",
       "CUT18451.1  CUT18451.1  H2B          gH2B    gH2B                           \n",
       "CUT18452.1  CUT18452.1  H2B          gH2B    gH2B                           \n",
       "\n",
       "           hgnc_gene_name taxonomy_id                      organism  \\\n",
       "CUT18449.1                    1473204  Lilium davidii var. unicolor   \n",
       "CUT18450.1                    1473204  Lilium davidii var. unicolor   \n",
       "CUT18445.1                    1473204  Lilium davidii var. unicolor   \n",
       "CUT18446.1                    1473204  Lilium davidii var. unicolor   \n",
       "CUT18447.1                    1473204  Lilium davidii var. unicolor   \n",
       "CUT18448.1                    1473204  Lilium davidii var. unicolor   \n",
       "CUT18451.1                    1473204  Lilium davidii var. unicolor   \n",
       "CUT18452.1                    1473204  Lilium davidii var. unicolor   \n",
       "\n",
       "                  phylum          class taxonomy_group info  \\\n",
       "CUT18449.1  Streptophyta  Magnoliopsida                       \n",
       "CUT18450.1  Streptophyta  Magnoliopsida                       \n",
       "CUT18445.1  Streptophyta  Magnoliopsida                       \n",
       "CUT18446.1  Streptophyta  Magnoliopsida                       \n",
       "CUT18447.1  Streptophyta  Magnoliopsida                       \n",
       "CUT18448.1  Streptophyta  Magnoliopsida                       \n",
       "CUT18451.1  Streptophyta  Magnoliopsida                       \n",
       "CUT18452.1  Streptophyta  Magnoliopsida                       \n",
       "\n",
       "                   references  \\\n",
       "CUT18449.1           26846354   \n",
       "CUT18450.1  26846354 32716939   \n",
       "CUT18445.1           26846354   \n",
       "CUT18446.1           26846354   \n",
       "CUT18447.1           26846354   \n",
       "CUT18448.1           26846354   \n",
       "CUT18451.1           26846354   \n",
       "CUT18452.1           26846354   \n",
       "\n",
       "                                                     sequence  \n",
       "CUT18449.1  MPPRRKKTAAGAAAGGKAAAAAVGKAGFMPPKKPKKGKKKTPIMRY...  \n",
       "CUT18450.1  MAPKKKPSKLVGTVTKTRKVTETQTLKVSLTKGLKPEDQQTTTNKF...  \n",
       "CUT18445.1  MAPKSEKKPAEKKPVAEKPAAEEEKKSAPAPAAAEKKPAEKKPKAG...  \n",
       "CUT18446.1  AEKKPKAGKKVPASKEGEKKKKRSKKSVETYKIYIFKVLKQVHPDI...  \n",
       "CUT18447.1  MAPKAEKKPAAKKPAATPPPEEEKEVVPPPPAEKKPKAGKKLPAAK...  \n",
       "CUT18448.1  AATPPPEEEKEVVPPPAEKKPAEKKPKAGKKLPASKEGDAKKKKKS...  \n",
       "CUT18451.1  MAPKAEKKPAAKKPAATPPPEEEKEVVPPPPAEKKPKAGKKLPAAK...  \n",
       "CUT18452.1      MAPKSEKKPAEKKPVAEKPAAEEEKKAAPAAAPAEKKAAEKKPKA  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cs.data[cs.data[\"variant\"] == \"gH2B\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sequence  self                                                      \n",
      "          other    MAPKAEKKPAAKKPAATPPPEEEKEVVPPPPAEKKPKAGKKLPAAK...\n",
      "Name: CUT18447.1, dtype: object\n",
      "sequence  self                                                      \n",
      "          other    AATPPPEEEKEVVPPPAEKKPAEKKPKAGKKLPASKEGDAKKKKKS...\n",
      "Name: CUT18448.1, dtype: object\n",
      "sequence  self                                                      \n",
      "          other    MAPKAEKKPAAKKPAATPPPEEEKEVVPPPPAEKKPKAGKKLPAAK...\n",
      "Name: CUT18451.1, dtype: object\n",
      "sequence  self                                                      \n",
      "          other    MPPRRKKTAAGAAAGGKAAAAAVGKAGFMPPKKPKKGKKKTPIMRY...\n",
      "Name: CUT18449.1, dtype: object\n",
      "sequence  self                                                      \n",
      "          other    MAPKSEKKPAEKKPVAEKPAAEEEKKSAPAPAAAEKKPAEKKPKAG...\n",
      "Name: CUT18445.1, dtype: object\n",
      "sequence  self                                                      \n",
      "          other    AEKKPKAGKKVPASKEGEKKKKRSKKSVETYKIYIFKVLKQVHPDI...\n",
      "Name: CUT18446.1, dtype: object\n",
      "sequence  self                                                  \n",
      "          other    MAPKSEKKPAEKKPVAEKPAAEEEKKAAPAAAPAEKKAAEKKPKA\n",
      "Name: CUT18452.1, dtype: object\n",
      "sequence  self                                                      \n",
      "          other    MAPKKKPSKLVGTVTKTRKVTETQTLKVSLTKGLKPEDQQTTTNKF...\n",
      "Name: CUT18450.1, dtype: object\n",
      "cp histones.csv backups/histones.csv-Mar0624163102\n",
      "Previous data backuped to backups/histones.csv-Mar0624163102\n",
      "Results saved to histones.csv\n"
     ]
    }
   ],
   "source": [
    "cs.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".conda-histdb_env",
   "language": "python",
   "name": "conda-env-.conda-histdb_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
